src/intro_gui.cpp

```diff
@@ -33,13 +39,152 @@
 
 #include "safeguards.h"
 
+
+struct IntroGameViewportCommand {
+	enum AlignmentH : byte {
+		LEFT,
+		CENTRE,
+		RIGHT,
+	};
+	enum AlignmentV : byte {
+		TOP,
+		MIDDLE,
+		BOTTOM,
+	};
+
+	int command_index;
+	Point position;
+	VehicleID vehicle;
+	uint delay;
+	int zoom_adjust;
+	bool pan_to_next;
+	AlignmentH align_h;
+	AlignmentV align_v;
+
+	IntroGameViewportCommand() : command_index(0), vehicle(INVALID_VEHICLE), delay(0), zoom_adjust(0), pan_to_next(false), align_h(CENTRE), align_v(MIDDLE)
+	{
+	}
+
+	Point PositionForViewport(const Viewport *vp)
+	{
+		if (this->vehicle != INVALID_VEHICLE) {
+			const Vehicle *v = Vehicle::Get(this->vehicle);
+			this->position = RemapCoords(v->x_pos, v->y_pos, v->z_pos);
+		}
+
+		Point p;
+		switch (this->align_h) {
+			case LEFT: p.x = this->position.x; break;
+			case CENTRE: p.x = this->position.x - vp->virtual_width / 2; break;
+			case RIGHT: p.x = this->position.x - vp->virtual_width; break;
+		}
+		switch (this->align_v) {
+			case TOP: p.y = this->position.y; break;
+			case MIDDLE: p.y = this->position.y - vp->virtual_height / 2; break;
+			case RIGHT: p.y = this->position.y - vp->virtual_height; break;
+		}
+		return p;
+	}
+};
+static std::vector<IntroGameViewportCommand> _intro_viewport_commands;
+
+
+void ReadIntroGameViewportCommands()
+{
+	_intro_viewport_commands.clear();
+
+	const char *sign_langauge = "^T\\s*([0-9]+)\\s*([-+TMBLCRP]+)\\s*([0-9]+)";
```

Comment:
`std::regex` has a reputation for being slow and horrible, and I'm not sure it's really necessary here - some minor simplifications of the format should make parsing/ignoring-on-error easier:
`T[0-9]+\s[-+TMBLCRP]+[0-9]+`

(still represented as regex because it's easier :p )

Reply:
I don't think speed is a concern, unless someone makes a title game with thousands of signs that would slow down the parsing.
I'm not sure what your simplification of the regex would improve. The spaces makes it more accepting of what authors find easy to write, and the parentheses are required for the match groups used in the parser.